ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/PushQueryResponseHandler.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.api.client.impl;
+
+import io.confluent.ksql.api.client.QueryResult;
+import io.confluent.ksql.api.client.Row;
+import io.confluent.ksql.api.server.protocol.QueryResponseMetadata;
+import io.confluent.ksql.rest.client.KsqlRestClientException;
+import io.vertx.core.Context;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.parsetools.RecordParser;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+
+public class PushQueryResponseHandler extends QueryResponseHandler<QueryResult> {
+
+  private QueryResultImpl queryResult;
+  private boolean paused;
+
+  PushQueryResponseHandler(final Context context, final RecordParser recordParser,
+      final CompletableFuture<QueryResult> cf) {
+    super(context, recordParser, cf);
+  }
+
+  @Override
+  protected void handleMetadata(final QueryResponseMetadata queryResponseMetadata) {
+    if (queryResponseMetadata.queryId == null || queryResponseMetadata.queryId.isEmpty()) {
+      cf.completeExceptionally(
+          new KsqlRestClientException("Expecting push query but was pull query"));
+      return;
+    }
+
+    this.queryResult = new QueryResultImpl(context, queryResponseMetadata.queryId,
+        Collections.unmodifiableList(queryResponseMetadata.columnNames),
+        Collections.unmodifiableList(queryResponseMetadata.columnTypes));
+    cf.complete(queryResult);
+  }
+
+  @Override
+  protected void handleRow(final Buffer buff) {
+    if (queryResult == null) {
+      throw new IllegalStateException("handleRow called before metadata processed");
+    }
+
+    final JsonArray values = new JsonArray(buff);
+    final Row row = new RowImpl(queryResult.columnNames(), queryResult.columnTypes(), values);
+    final boolean full = queryResult.accept(row);
+    if (full && !paused) {
+      recordParser.pause();
+      queryResult.drainHandler(this::publisherReceptive);
+      paused = true;
+    }
+  }
+
+  @Override
+  protected void handleBodyEnd() {
+  }
+
+  private void publisherReceptive() {
+    paused = false;
```

Comment:
Maybe a checkContext here just to be sure?

Reply:
Added. I don't fully understand what the purpose of these context checks is, though. (I've seen them throughout the server code as well but haven't fully wrapped my head around them.) What exactly are we trying to prevent?

Reply:
They're not strictly necessary but I like to add them as sanity checks.
Basically they ensure that the particular method is being executed on the right thread.
One nice thing about the Vert.x threading model is it ensures that all handlers when set from a specific context are called on the same context too, i.e. usually that means on the same thread. That means you don't have to worry about synchronizing stuff, race conditions, deadlocks etc. Big win! However... maybe there's some other thread in your code that is mistakenly calling that method too, in which case it might silently be corrupting the state as that state is not synchronized. The checkContext calls check this isn't occurring.

Reply:
Super helpful explanation. Thanks!